README.md
# flutter_lib_admob

a flutter lib utils for admob.

## How To Use

### 示例Demo
你可以在这里看到示例Demo:[admob_demo](https://github.com/MeDeity/admob_demo)
![demo_download_address](images/download_url.png)


### 平台相关
#### 一、Android 配置:
在AndroidManifest.xml中配置AppId(以下以官方测试用ID为例):
```
<application>
...
<meta-data
    android:name="com.google.android.gms.ads.APPLICATION_ID"
    android:value="ca-app-pub-3940256099942544~3347511713"/>
...
</application>
```

#### 二、IOS配置
在Info.plist中加入以下内容
```
<key>GADApplicationIdentifier</key>
<string>ca-app-pub-3940256099942544~1458002511</string>
```

#### 三、初始化
```
AdmobUtils.initAdmobIds(
      appIdList: [
        "ca-app-pub-3940256099942544~3347511713",
        "ca-app-pub-3940256099942544~1458002511"
      ],
      androidBannerIdList:["ca-app-pub-3940256099942544/6300978111"],
      iosBannerIdList:["ca-app-pub-3940256099942544/2934735716"],
      androidInterstitialIdList:["ca-app-pub-3940256099942544/1033173712"],
      iosInterstitialIdList:["ca-app-pub-3940256099942544/4411468910"],
      androidRewardBasedVideoIdList:["ca-app-pub-3940256099942544/5224354917"],
      iosRewardBasedVideoIdList:["ca-app-pub-3940256099942544/1712485313"]
    );
///add init and listener callback    
AdmobUtils.initData((AdmobAdEvent event, Map<String, dynamic> args, AdTypeEnum adType) {
      print("${adType.index}");
    });    
///init admob config    
GenerateAdmobId factory = AdmobUtils.getAdmobIdFactory();
Admob.initialize(factory.getAppId());
```

#### 四、展示广告

```dart
///展示插页式AD
AdmobUtils.showInterstitialAd();
///展示奖励式AD
AdmobUtils.showRewardAd();
///retrun a widget
Widget banner = AdmobUtils.showBannerAd(
    callback: (AdmobAdEvent event, Map<String, dynamic> args,AdTypeEnum adType) {
    ///do something
}),
```




CHANGELOG.md
## [0.0.1] - 2019-10-23.

* first release at 2019-10-23 18:03:18.


test/flutter_lib_admob_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_lib_admob/flutter_lib_admob.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/admob/admob_utils.dart
import 'package:admob_flutter/admob_flutter.dart';
import 'package:flutter/material.dart';
import 'package:flutter_lib_admob/admob/ad_type_enum.dart';
import 'package:flutter_lib_admob/admob/generate_admob_id.dart';

///
/// 工具类汇总
/// create by fengwenhua at 2019-10-21 17:22:11
///
class AdmobUtils {
  static GenerateAdmobId _generateAdmobId;
  static AdmobInterstitial interstitialAd;
  static AdmobReward rewardAd;

  ///初始化id
  static void initAdmobIds({
    List<String> appIdList,
    List<String> androidBannerIdList,
    List<String> iosBannerIdList,
    List<String> androidInterstitialIdList,
    List<String> iosInterstitialIdList,
    List<String> androidRewardBasedVideoIdList,
    List<String> iosRewardBasedVideoIdList}) {
    _generateAdmobId = GenerateAdmobId(appIdList: appIdList,
        androidBannerIdList: androidBannerIdList,
        iosBannerIdList: iosBannerIdList,
        androidInterstitialIdList: androidInterstitialIdList,
        iosInterstitialIdList: iosInterstitialIdList,
        androidRewardBasedVideoIdList: androidRewardBasedVideoIdList,
        iosRewardBasedVideoIdList: iosRewardBasedVideoIdList);
  }

  static GenerateAdmobId getAdmobIdFactory(){
    return _generateAdmobId;
  }


  ///回调处理
  static void initData(EventCallback callback) {
    interstitialAd = AdmobInterstitial(
      adUnitId: _generateAdmobId.getInterstitialRandomId(),
      listener: (AdmobAdEvent event, Map<String, dynamic> args) {
        if (event == AdmobAdEvent.closed) interstitialAd.load();
        callback(event, args, AdTypeEnum.Interstitial);
      },
    );

    rewardAd = AdmobReward(
        adUnitId: _generateAdmobId.getRewardRandomId(),
        listener: (AdmobAdEvent event, Map<String, dynamic> args) {
          if (event == AdmobAdEvent.closed) rewardAd.load();
          callback(event, args, AdTypeEnum.Reward);
        });

    interstitialAd.load();
    rewardAd.load();
  }

  static void showInterstitialAd() {
    interstitialAd?.show();
  }

  static void showRewardAd() {
    rewardAd?.show();
  }

  static Widget showBannerAd({AdmobBannerSize bannerSize:AdmobBannerSize.BANNER,EventCallback callback}) {
    return AdmobBanner(
      adUnitId: _generateAdmobId.getBannerRandomId(),
      adSize: bannerSize,
      listener:(AdmobAdEvent event, Map<String, dynamic> args) {
        callback(event, args, AdTypeEnum.Reward);
      },
    );
  }

  static void disposeInterstitialAd() {
    interstitialAd?.show();
  }

  static void disposeRewardAd() {
    rewardAd?.show();
  }


  ///测试用方法
  static void initTestData(EventCallback callback) {
    _generateAdmobId = GenerateAdmobId()
      ..initTestIds();

    interstitialAd = AdmobInterstitial(
      adUnitId: _generateAdmobId.getInterstitialRandomId(),
      listener: (AdmobAdEvent event, Map<String, dynamic> args) {
        callback(event, args, AdTypeEnum.Interstitial);
      },
    );

    rewardAd = AdmobReward(
        adUnitId: _generateAdmobId.getRewardRandomId(),
        listener: (AdmobAdEvent event, Map<String, dynamic> args) {
          if (event == AdmobAdEvent.closed) rewardAd.load();
          callback(event, args, AdTypeEnum.Reward);
        });

    interstitialAd.load();
    rewardAd.load();
  }
}

typedef void EventCallback(AdmobAdEvent event, Map<String, dynamic> args,
    AdTypeEnum adType);


lib/admob/generate_admob_id.dart
import 'dart:io';
import 'dart:math';

import 'package:flutter/material.dart';

///
///admob id 生成器
///create by fengwenhua at 2019-10-21 14:44:52
///
class GenerateAdmobId {
  ///存放appIds(android/ios)
  List<String> appIdList;

  ///存放bannerIds(android/ios) --Banner
  List<String> androidBannerIdList = List();
  List<String> iosBannerIdList = List();

  ///存放interstitialIds(android/ios)-插页式
  List<String> androidInterstitialIdList = List();
  List<String> iosInterstitialIdList = List();

  ///存放RewardBasedVideoIds(android/ios)- 奖励式Video
  List<String> androidRewardBasedVideoIdList = List();
  List<String> iosRewardBasedVideoIdList = List();

  GenerateAdmobId(
      {this.appIdList,
      this.androidBannerIdList,
      this.iosBannerIdList,
      this.androidInterstitialIdList,
      this.iosInterstitialIdList,
      this.androidRewardBasedVideoIdList,
      this.iosRewardBasedVideoIdList});

  ///返回随机Id
  String getRandomIds(List<String> list) {
    String result = "";
    if (null != list && list.isNotEmpty) {
      int size = list.length;
      int seed = Random().nextInt(size);
      result = list[seed];
    }
    return result;
  }

  ///返回app id
  String getAppId(){
    String result = "";
    int size = appIdList?.length??0;
    if(Platform.isAndroid){
      if(size>=1){
        result = appIdList[0];
      }
    }else if(Platform.isIOS){
      if(size>=2){
        result = appIdList[1];
      }
    }
    return result;
  }

  ///返回banner id
  String getBannerRandomId(){
    String result = "";
    if(Platform.isAndroid){
      result = getRandomIds(androidBannerIdList);
    }else if(Platform.isIOS){
      result = getRandomIds(iosBannerIdList);
    }
    return result;
  }


  ///返回Interstitial id
  String getInterstitialRandomId(){
    String result = "";
    if(Platform.isAndroid){
      result = getRandomIds(androidInterstitialIdList);
    }else if(Platform.isIOS){
      result = getRandomIds(iosInterstitialIdList);
    }
    return result;
  }

  ///返回Reward id
  String getRewardRandomId(){
    String result = "";
    if(Platform.isAndroid){
      result = getRandomIds(androidRewardBasedVideoIdList);
    }else if(Platform.isIOS){
      result = getRandomIds(iosRewardBasedVideoIdList);
    }
    return result;
  }

  //
  //Test Id's from: 测试用的ID值
  //https://developers.google.com/admob/ios/banner
  //https://developers.google.com/admob/android/banner
  //
  //App Id
  //Android: ca-app-pub-3940256099942544~3347511713
  //iOS: ca-app-pub-3940256099942544~1458002511
  //
  //Banner
  //Android: ca-app-pub-3940256099942544/6300978111
  //iOS: ca-app-pub-3940256099942544/2934735716
  //
  //Interstitial
  //Android: ca-app-pub-3940256099942544/1033173712
  //iOS: ca-app-pub-3940256099942544/4411468910
  //
  //Reward Video
  //Android: ca-app-pub-3940256099942544/5224354917
  //iOS: ca-app-pub-3940256099942544/1712485313
  //
  void initTestIds() {
    this.appIdList = [
      "ca-app-pub-3940256099942544~3347511713",
      "ca-app-pub-3940256099942544~1458002511"
    ];

    ///android+ios app id;
    this.androidBannerIdList = ["ca-app-pub-3940256099942544/6300978111"];

    ///android banner id
    this.iosBannerIdList = ["ca-app-pub-3940256099942544/2934735716"];

    ///Interstitial id
    this.androidInterstitialIdList = ["ca-app-pub-3940256099942544/1033173712"];
    this.iosInterstitialIdList = ["ca-app-pub-3940256099942544/4411468910"];

    ///Reward Video
    this.androidRewardBasedVideoIdList = [
      "ca-app-pub-3940256099942544/5224354917"
    ];
    this.iosRewardBasedVideoIdList = ["ca-app-pub-3940256099942544/1712485313"];
  }
}
